llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -936,13 +937,83 @@ Error MCCASBuilder::createStringSection(
   return Error::success();
 }
 
+/// Reads and returns the length field of a debug info header contained in
+/// Reader, assuming Reader is positioned at the beginning of the section.
+/// The Reader's state is advanced to the first byte after the section.
+Expected<size_t> getSizeFromDebugHeaderAndSkip(BinaryStreamReader &Reader) {
+  // From Dwarf 5 section 7.4:
+  // In the 32-bit DWARF format, an initial length field [...] is an unsigned
+  // 4-byte integer (which must be less than 0xfffffff0);
+  uint32_t Word1;
+  if (auto E = Reader.readInteger(Word1))
+    return E;
+
+  // TODO: handle 64-bit DWARF format.
+  assert(Word1 < 0xfffffff0);
+
+  if (auto E = Reader.skip(Word1))
+    return E;
+  return Word1;
+}
+
+/// For each compile unit inside the __debug_info section, create a
+/// corresponding DebugInfoRef CAS object.
+Error MCCASBuilder::splitDebugInfoCUs() {
+  auto *DebugInfoSection =
+      Asm.getContext().getObjectFileInfo()->getDwarfInfoSection();
+  if (!DebugInfoSection)
+    return Error::success();
+
+  const auto &FragmentList = DebugInfoSection->getFragmentList();
+  assert(FragmentList.size() == 1 && "One fragment in debug str section");
+
+  startSection(DebugInfoSection);
+
+  // Store relocation information into the MCCASBuilder's state.
+  const auto &DebugFragment = cast<MCDataFragment>(FragmentList.front());
+  auto RelocIt = RelMap.find(&DebugFragment);
+  if (RelocIt != RelMap.end())
+    SectionRelocs.append(RelocIt->second.begin(), RelocIt->second.end());
+
+  ArrayRef<char> DebugInfoData = DebugFragment.getContents();
+  StringRef DebugInfoStrRef(DebugInfoData.data(), DebugInfoData.size());
+  auto Reader = BinaryStreamReader(DebugInfoStrRef, Asm.getBackend().Endian);
+
+  // CU splitting loop.
+  while (!Reader.empty()) {
+    auto StartOffset = Reader.getOffset();
+    auto CULength = getSizeFromDebugHeaderAndSkip(Reader);
+    if (!CULength)
+      return CULength.takeError();
+
+    auto CUData =
```

Comment:
if this is a `StringRef`, I would just declare it as such and reserve `auto` for unergonomically long or uninteresting types like iterators.

Reply:
(same for the other auto variables)

Reply:
Oops, my bad. I'm coming from a codebase with the "Almost Always Auto" policy, please bear with me as I get used to reverting this mindset 😅

Reply:
I'm not actually sure if that's my personal preference or an actual LLVM style thing, but it does make the code a bit easier to read/review if the (sane) types are explicit.

Reply:
Oh it is very much an LLVM requirement to _not_ use auto everywhere. As you said, the programmer guide essentially prescribes using `auto` for the crazy types and also if the RHS is a cast or something that already spells out the type.

We made an exception to it in my previous project because... well, someone named Felipe wrote the coding guidelines for that project 👀 We should talk about this over coffee some day, in my experience everyone has the same initial reaction that you do, but over time they change their opinion once they use this style on a daily basis. The style brings a lot of cool advantages both in how it forces one to write code, but also in how to just flat out forbid anti-patterns... Look at me rambling when I said this is better as a coffee time conversation.